op-node/rollup/driver/sequencer.go

```diff
@@ -210,29 +203,34 @@ func (d *Sequencer) startBuildingEspressoBatch(ctx context.Context, l2Head eth.L
 		NmtProofs: nil,
 	}
 	d.espressoBatch = batch
-	return d.updateEspressoBatch(ctx, blocks)
+	return d.updateEspressoBatch(ctx, blocks.Window, blocks.Next)
 }
 
 // updateEspressoBatch appends the transactions contained in the Espresso blocks denoted by
-// `newHeaders` to the current in-progress batch. If the batch is complete (Espresso has sequenced
-// at least one block with a timestamp beyond the end of the current sequencing window) it will set
-// the `complete` flag.
-func (d *Sequencer) updateEspressoBatch(ctx context.Context, newHeaders []espresso.Header) error {
+// `newHeaders` to the current in-progress batch. If `end`, the first block after the window of this
+// batch, is available, it will be saved in the `NextBatchFirstBlock` field of the batch
+// justification.
+func (d *Sequencer) updateEspressoBatch(ctx context.Context, newHeaders []espresso.Header, end *espresso.Header) error {
 	batch := d.espressoBatch
 	for i := range newHeaders {
-		if newHeaders[i].Timestamp >= batch.windowEnd {
-			batch.jst.Payload.NextBatchFirstBlock = newHeaders[i]
-			return nil
+		header := &newHeaders[i]
+
+		// Validate that the given header is in the window.
+		if header.Timestamp >= batch.windowEnd {
```

Comment:
Shouldn't we also check that header.Timestamp >= batch.windowStart?

Reply:
Oh, yes. And in particular >= the last header